shakytweaks/src/main/java/com/mindsea/shakytweaks/ShakyTweaks.kt

```diff
@@ -44,18 +46,21 @@ object ShakyTweaks {
 
     private const val KEY_PRESS_THRESHOLD = 50L
 
-    fun init(context: Context) {
+    fun init(activity: Activity, lifecycle: Lifecycle) {
```

Comment:
Hey I think is not a good ideia pass an activity reference to a library.
This activity can become null.

and this line can crash the app.
activity.startActivity(createTweaksActivityIntent(activity))

In some case you can use the application to access android resources.

Reply:
I can't understand your point. Could you provide an example?

Reply:
Of course!

When you do some thing like this:
ShakyTweaks.init(**applicationContext**)   the context is always not null, because is a application reference.

When you do it:
ShakyTweaks.init(**this@MainActivity**, lifecycle) the context can be null because your activity can be destroyed.

if you library try to use the activity reference, in some moment the library can get a NullPointerException.

Reply:
You can see a exemplo like Koin library:

```
class MyApplication : Application() {
  override fun onCreate(){
    super.onCreate()
    // start Koin!
    startKoin {
      // declare used Android context
      androidContext(this@MyApplication)
      // declare modules
      modules(myModule)
    }
  } 
} 
```

Reply:
@romulojjunior Okay, I got it now.

I think the activity reference is not invalid in this case because we have the following condition:
`
if(lifecycle.currentState.isAtLeast(Lifecycle.State.RESUMED)) {
`

If the state is at least `RESUMED`, the _activity_ has a valid reference to launch another one.

Another aspect is that using this approach, We can avoid the issue opening activities from the base context in Android 10.

Reply:
The Koin library don't open any Activity. In our case we need to handle that : (

Reply:
Well, you can try to use your solution.

Reply:
I guess your activity won't leak, because you are calling `release` on Activity's `onDestroy`.

But that is a problem:
Now, users of this library MUST use `ShakyTweaks` release on all `onDestroy` of their activities. If a user of the library forgets, there will be bugs. IMO this is a problem, because we can't rely on user reading the documentation carefully to make things right. If there are means to restrict erroneous code by forms of better lib architecture, I would go for that.

Also, the context type doesn't necessarily needs to be of Activity type. The could would work if it was an application context. I.E: `startActivity` would work using application context, because TweaksActivity sue the correct intent flags for a new activity.


Maybe your solution would work, but it is very intrusive. Imagine the apps that already use this lib, they would have to go adding changes in all activities of their code, possibly forgetting some steps and causing bugs 🤯 .

Try to see if there is a way to solve this, only using the context from the application 🤷 . Or try to arrange this in a way that clients of this lib can use it easier and safely.

Reply:
@gianei You are right, it is dangerous. The problem that I don't have any idea how to solve this :( . Maybe another strategy would be we create an abstract activity, something like `ShakyTweaksActivity`. In this activity, we can handle the behavior in `onCreate` and `onDestroy` methods. The dev user just needs to extends from that activity. What do you think about it?

Reply:
Maybe, but it is still restrictive. Inheritance is also a bit complicated, now we are restricting all client's activities to extend our ShakyTweaks activity, which extends AppCompatActivity. But what if someone wants to build something that does not extend AppCompatActivity? I am not sure... seems bad hehe.

Take a deeper look in the [Lifecycle thing here](https://developer.android.com/topic/libraries/architecture/lifecycle#lco).
I think you can start a new LifecycleAware component, in a way that you don't need call `release`... and use
`@OnLifecycleEvent(Lifecycle.Event.ON_STOP)` to stop listening to shakes

Reply:
Pardon the Android Noob here. But would it be a bad idea to require the user of ShakeyTweaks to tell the library about relevant lifecycle events?
That way, the user can basically enable/disable the tweaks in whatever way fits their situation.

All we then have to do is add, say, 2 new functions and update the readme.

Reply:
I think this is just what Gabriel did. It does work, but IMO it makes the library harder to use.
I hope there is other better approaches, but if we can't find anything better, than we have to consider the drawbacks...

Reply:
@gianei / @romulojjunior / @hofmanntm I updated the code. I tried to follow the comments that you made ;)